client/hotrod-client/src/test/java/org/infinispan/client/hotrod/StorageRoutingTest.java

```diff
@@ -0,0 +1,115 @@
+import org.infinispan.configuration.cache.StorageType;
+import org.infinispan.server.hotrod.HotRodServer;
+import org.testng.annotations.Test;
+
+/**
+ * Tests that the Hot Rod client can correctly route requests to a server using different {@link StorageType}.
+ *
+ * @since 11.0
+ */
+@Test(groups = "functional", testName = "client.hotrod.StorageRoutingTest")
+public class StorageRoutingTest extends MultiHotRodServersTest {
+
+   private static final int CLUSTER_SIZE = 3;
+
+   protected StorageType storageType;
+   private Object key;
+
+   public Object[] factory() {
+      String stringKey = "key";
+      byte[] byteArrayKey = new byte[]{1, 2, 3};
+
+      return new Object[]{
+            new StorageRoutingTest().withStorageType(OBJECT).withKey(stringKey),
+            new StorageRoutingTest().withStorageType(OBJECT).withKey(byteArrayKey),
+            new StorageRoutingTest().withStorageType(OFF_HEAP).withKey(stringKey),
+            new StorageRoutingTest().withStorageType(OFF_HEAP).withKey(byteArrayKey),
+            new StorageRoutingTest().withStorageType(BINARY).withKey(stringKey),
+            new StorageRoutingTest().withStorageType(BINARY).withKey(byteArrayKey)
+      };
+   }
+
+   protected String[] parameterNames() {
+      return new String[]{"storage", "key"};
+   }
+
+   protected Object[] parameterValues() {
+      return new Object[]{storageType, key};
+   }
+
+   private StorageRoutingTest withStorageType(StorageType storageType) {
+      this.storageType = storageType;
+      return this;
+   }
+
+   private StorageRoutingTest withKey(Object key) {
+      this.key = key;
+      return this;
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      ConfigurationBuilder cfgBuilder = getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, false);
+      cfgBuilder.jmxStatistics().enable();
+      cfgBuilder.clustering().hash().numOwners(1);
+      cfgBuilder.memory().storageType(storageType);
+      createHotRodServers(CLUSTER_SIZE, cfgBuilder);
+      waitForClusterToForm();
+   }
+
+   @Override
+   protected org.infinispan.client.hotrod.configuration.ConfigurationBuilder createHotRodClientConfigurationBuilder(HotRodServer hotRodServer) {
+      org.infinispan.client.hotrod.configuration.ConfigurationBuilder hotRodClientConfigurationBuilder = super.createHotRodClientConfigurationBuilder(hotRodServer);
+      hotRodClientConfigurationBuilder.balancingStrategy(() -> new FixedServerBalancing(hotRodServer));
+      return hotRodClientConfigurationBuilder;
+   }
+
+   @Test
+   public void shouldContactKeyOwnerForPutGet() {
+      String value = "value";
+      RemoteCache<Object, String> remoteCache = clients.get(0).getCache();
+      remoteCache.put(key, value);
+
+      assertEquals(remoteCache.get(key), "value");
+
+      assertCorrectServerContacted();
+   }
+
+   private void assertCorrectServerContacted() {
+      AtomicInteger storedIn = new AtomicInteger(-1);
```

Comment:
any reason to use `AtomicInteger`? a `boolean` would be enough IMO :)

Reply:
conciseness only

Reply:
ah! I missed the last assert `assertEquals(storeServer, retrieveServer, ...)` just ignore my comment.